### PR TITLE
KAFKA-15458: Fully resolve endpoint information before registering controllers

### PR DESCRIPTION
**Pull request by cmccabe**
Endpoint information provided by KafkaConfig can be incomplete in two ways. One is that endpoints using ephemeral ports will show up as using port 0. Another is that endpoints binding to 0.0.0.0 will show up with a null or blank hostname. Because we were not accounting for this in controller registration, it was leading to a null pointer dereference when we tried to register a controller using an endpoint defined as PLAINTEXT://:9092.

This PR adds a ListenerInfo class which can fix both of the causes of incomplete endpoint information. It also handles serialization to and from various RPC and record formats. This allows us to remove a lot of boilerplate code and standardize the handling of listeners between BrokerServer and ControllerServer.

### CODE REVIEW

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+ * different data structures, and also to handle the two big gotchas of Kafka endpoints.
+ *
+ * The first gotcha is that the hostname will be null or blank if we are listening on 0.0.0.0.
+ * The withWildcardHostnamesResolved function creates a ListenerInfo object where all such hostnames
+ * are replaced by specific hostnames. (It's not perfect because we have to choose a single hostname
+ * out of multiple possibilities. In production scenarios it would be better to set the desired
+ * hostname explicitly in the configuration rather than binding to 0.0.0.0.)
+ *
+ * The second gotcha is that if someone configures an ephemeral port (aka port 0), we need to fill
+ * The withEphemeralPortsCorrected resolves this by filling in the missing information for ephemeral
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java]: seems some words are missing here?

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+    public static ListenerInfo create(
+        Optional<String> firstListenerName,
+        List<Endpoint> rawListeners
+    ) {
+        LinkedHashMap<String, Endpoint> listeners = new LinkedHashMap<>();
+        for (Endpoint listener : rawListeners) {
+            String name = listener.listenerName().get();
+            if (Optional.of(name).equals(firstListenerName)) {
+                listeners.put(name, listener);
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java:192]: Can probably break here if you find the first listener.

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    private ListenerInfo(Map<String, Endpoint> listeners) {
+        this.listeners = Collections.unmodifiableMap(listeners);
+    }
+
+    public Map<String, Endpoint> listeners() {
+        return listeners;
+    }
+
+    public Endpoint firstListener() {
+        return listeners.values().iterator().next();
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java:223]: Is the iterator guaranteed to be non-empty?

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+     * Create a new ListenerInfo object where null or blank hostnames (signifying that the user
+     * asked to bind to 0.0.0.0) are replaced by specific hostnames.
+     *
+     * @return A new ListenerInfo object.
+     */
+    public ListenerInfo withWildcardHostnamesResolved() throws UnknownHostException {
+        LinkedHashMap<String, Endpoint> newListeners = new LinkedHashMap<>();
+        for (Map.Entry<String, Endpoint> entry : listeners.entrySet()) {
+            if (entry.getValue().host() == null || entry.getValue().host().trim().isEmpty()) {
+                String newHost = InetAddress.getLocalHost().getCanonicalHostName();
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java:236]: Should we log something to the user if we are resolving localhost as the host name? As you mention in the class javadoc, normally for production use-cases we expect the hostname to be explicitly set.

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+     * @param getBoundPortCallback  The callback used to correct ephemeral endpoints.
+     *
+     * @return A new ListenerInfo object.
+     */
+    public ListenerInfo withEphemeralPortsCorrected(Function<String, Integer> getBoundPortCallback) {
+        LinkedHashMap<String, Endpoint> newListeners = new LinkedHashMap<>();
+        for (Map.Entry<String, Endpoint> entry : listeners.entrySet()) {
+            if (entry.getValue().port() == 0) {
+                Endpoint prevEndpoint = entry.getValue();
+                int newPort = getBoundPortCallback.apply(entry.getKey());
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java:268]: Same logging question as above

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+
+    private static void checkPortIsSerializable(int port) {
+        if (port == 0) {
+            throw new RuntimeException("Cannot serialize ephemeral port 0 in ListenerInfo.");
+        } else if (port < 0) {
+            throw new RuntimeException("Cannot serialize negative port number " + port +
+                    " in ListenerInfo.");
+        } else if (port > 65535) {
+            throw new RuntimeException("Cannot serialize invalid port number " + port +
+                    " in ListenerInfo.");
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java]: nit: too much indentation

File: metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java (hunk 3 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/metadata/ListenerInfo.java:223]: In general the cases where we have no listeners in here are somewhat degenerate (unit tests with garbage registrations, etc.) and we don't call firstListener in `QuorumController` code anyway.

However, I'll add a check and a `throw` to be more explicit.

**Review by mumrah — APPROVED** [review]
Thanks for the updates @cmccabe. There are several failing tests including `ControllerRegistrationManagerTest`. Not sure if it's related to this PR.

This LGTM pending a reasonable Jenkins run

**Comment by jolshan** [comment]
Why did we merge this without checking the ControllerRegistrationManagerTest? This is not a reasonable run: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14376/3/#showFailuresLink

**Pull request merged**